subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -0,0 +1,1295 @@
+			goto clean_up;
+		}
+
+		ret = do_api_call(&http_req, rest_ctx, HTTPS_PORT, CONFIG_NRF_CLOUD_SEC_TAG);
+		if (ret) {
+			ret = -EIO;
+			goto clean_up;
+		}
+
+		if (rest_ctx->status != HTTP_STATUS_PARTIAL) {
+			ret = -EBADMSG;
+			goto clean_up;
+		}
+
+		if (total_bytes == 0) {
+			total_bytes = get_content_range_total_bytes(rest_ctx->rx_buf);
+			if (total_bytes <= 0) {
+				ret = -EBADMSG;
+				goto clean_up;
+			}
+			LOG_DBG("Total bytes in payload: %d", total_bytes);
+
+			if (!result) {
+				/* If all data was able to be downloaded without
+				 * a result buffer, return 0.  Otherwise, return
+				 * the total bytes needed for the result.
+				 */
+				if (total_bytes > frag_size) {
+					ret = total_bytes;
+				} else {
+					ret = 0;
+				}
+
+				goto clean_up;
+
+			} else if (result->buf_sz < total_bytes) {
+				LOG_ERR("Provided buffer is too small");
+				ret = -ENOBUFS;
+			}
+		}
+
+		rcvd_bytes += rest_ctx->response_len;
+
+		LOG_DBG("AGPS data rx: %u/%u", rcvd_bytes, total_bytes);
+		if (rcvd_bytes > total_bytes) {
+			ret = -EFBIG;
+			goto clean_up;
+		}
+
+		memcpy(&result->buf[pos],
+		       rest_ctx->response,
+		       rest_ctx->response_len);
+
+		pos += rest_ctx->response_len;
+		remain = total_bytes - rcvd_bytes;
+
+	} while (remain);
+
+	/* Set output size */
+	result->agps_sz = total_bytes;
+
+clean_up:
+	if (url) {
+		k_free(url);
+	}
+	if (auth_hdr) {
+		k_free(auth_hdr);
+	}
+
+	close_connection(rest_ctx);
+
+	return ret;
+}
+
+int nrf_cloud_rest_pgps_data_get(struct nrf_cloud_rest_context *const rest_ctx,
+				 struct nrf_cloud_rest_pgps_request const *const request,
+				 struct nrf_cloud_pgps_result *const result)
+{
+	__ASSERT_NO_MSG(rest_ctx != NULL);
+	__ASSERT_NO_MSG(request != NULL);
+	__ASSERT_NO_MSG(request->device_id != NULL);
+	/* result is not required, user can parse the response if they wish */
+
+	int ret;
+	size_t url_sz;
+	size_t remain;
+	size_t pos;
+	char *auth_hdr = NULL;
+	char *url = NULL;
+	struct http_request http_req;
+
+	init_request(&http_req, HTTP_GET, CONTENT_TYPE_ALL);
+
+	/* Determine size of URL buffer and allocate */
+	url_sz = sizeof(API_GET_PGPS_BASE) + strlen(request->device_id);
+
+	if (request->pgps_req) {
+		if (request->pgps_req->prediction_count !=
+		NRF_CLOUD_REST_PGPS_REQ_NO_COUNT) {
+			url_sz += sizeof(PGPS_REQ_PREDICT_CNT) +
+				  UINT32_MAX_STR_SZ;
+		}
+		if (request->pgps_req->prediction_period_min !=
+		    NRF_CLOUD_REST_PGPS_REQ_NO_INTERVAL) {
+			url_sz += sizeof(PGPS_REQ_PREDICT_INT_MIN) +
+				  UINT32_MAX_STR_SZ;
+		}
+		if (request->pgps_req->gps_day !=
+		    NRF_CLOUD_REST_PGPS_REQ_NO_GPS_DAY) {
+			url_sz += sizeof(PGPS_REQ_START_GPS_DAY) +
+				  UINT32_MAX_STR_SZ;
+		}
+		if (request->pgps_req->prediction_count !=
+		    NRF_CLOUD_REST_PGPS_REQ_NO_GPS_TOD) {
+			url_sz += sizeof(PGPS_REQ_START_GPS_TOD_S) +
+				  UINT32_MAX_STR_SZ;
+		}
+	}
+
+	url = k_calloc(url_sz, 1);
+	if (!url) {
+		ret = -ENOMEM;
+		goto clean_up;
```

Comment:
I think returning is safe here, same in the A-GPS request API

Reply:
the only issue with returning here is that `close_connection` will not be called.
so if a user has already established a connection and has set the the keep alive flag to false, the connection should be closed.
maybe that flag is a bit silly... another option would be to remove the flag, always leave the connection open, and require the user to call `nrf_cloud_rest_disconnect` when they want to close?

Reply:
Hm, I see.  I noticed there were some early returns in `nrf_cloud_rest_agps_data_get()`, so didn't think of that. Are those returns okay?
I think the flag can be useful to avoid (temporary) resource leaks, so it could be worth keeping. I don't know for how long the connection will be kept open if a user forgets to call `nrf_cloud_rest_disconnect ()`, but I assume it may take a while. During that time, resources in the modem are tied up and for instance opening a new TLS connection is likely to fail (assuming one REST connection + one cloud connection using TLS are open).

Reply:
good catch, i have removed those early returns.